zkevm-circuits/src/pi_circuit2.rs

```diff
@@ -380,1147 +220,479 @@ impl<F: Field> SubCircuitConfig<F> for PiCircuitConfig<F> {
-                            is_final,
-                            gas_cost,
-                            &mut rpi_vals,
-                        )?;
-                        offset += 1;
-                        calldata_count += 1;
-                    }
-                }
-                for _ in calldata_count..self.max_calldata {
-                    self.assign_tx_calldata_row(
-                        &mut region,
-                        offset,
-                        0, // tx_id
-                        0,
-                        0,
-                        F::zero(),
+                let mut rpi_rlc_acc_cell = None;
+                for (annotation, field_type, field_bytes) in public_data.assignments() {
+                    let cells = self.assign_pi_field(
+                        region,
+                        &mut offset,
+                        annotation,
+                        &field_bytes,
+                        &mut rpi_rlc_acc,
+                        &mut rpi_len_acc,
+                        challenges,
                         false,
-                        F::zero(),
-                        &mut rpi_vals,
                     )?;
-                    offset += 1;
+                    if field_type == FieldType::BlockHash {
+                        region.constrain_equal(
+                            block_table_hash_cell.cell(),
+                            cells[RPI_CELL_IDX].cell(),
+                        )?;
+                    }
+                    rpi_rlc_acc_cell = Some(cells[RPI_RLC_ACC_CELL_IDX].clone());
                 }
-                // NOTE: we add this empty row so as to pass mock prover's check
-                //      otherwise it will emit CellNotAssigned Error
-                let tx_table_len = TX_LEN * self.max_txs + 1;
-                self.assign_tx_empty_row(&mut region, tx_table_len + offset)?;
-
-                let (origin_txs_hash_hi_cell, origin_txs_hash_lo_cell, txs_rlc_acc) =
-                    self.assign_txs(&mut region, public_data, challenges, rpi_vals)?;
-                // assert two txs hash are equal
-                region.constrain_equal(txs_hash_hi_cell.cell(), origin_txs_hash_hi_cell.cell())?;
-                region.constrain_equal(txs_hash_lo_cell.cell(), origin_txs_hash_lo_cell.cell())?;
-                Ok((
-                    [block_hash_hi_cell, block_hash_lo_cell],
-                    txs_rlc_acc,
-                    block_rlc_acc,
-                ))
+
+                // input_rlc in self.rpi
+                // input_len in self.rpi_len_acc
+                // output_rlc in self.rpi_rlc_acc
+                let keccak_row = offset;
+                let rpi_rlc_acc_cell = rpi_rlc_acc_cell.unwrap();
+                rpi_rlc_acc_cell.copy_advice(
+                    || "keccak(rpi)_input",
+                    region,
+                    self.rpi,
+                    keccak_row,
+                )?;
```

Comment:
Seems like this logic with the cells and stuff can can be simplified a bit because hashing is always done at the same position, so just enable the keccak lookup on the last row which already contains all the necessary data and good to go?

Reply:
I need use two cell to store the `keccak_input_rlc` and `keccak_output_rlc`, so I reuse the `self.rpi` and `self.rpi_rlc_acc` which previously used to store `rpi` and `rpi_rlc`, the `rpi_rlc` here as the `keccak_input_rlc` for keccak lookup
So, if I don't copy from the `rpi_rlc` column into the `rpi` column, I need allocate an new column for `keccak_output_rlc`